src/main/java/org/springframework/data/couchbase/core/convert/MappingCouchbaseConverter.java

```diff
@@ -959,10 +947,190 @@ public ConverterAwareSpELExpressionParameterValueProvider(final SpELExpressionEv
 			this.parent = parent;
 		}
 
-		@Override
-		protected <T> T potentiallyConvertSpelValue(final Object object,
+		@Override protected <T> T potentiallyConvertSpelValue(final Object object,
 				final Parameter<T, CouchbasePersistentProperty> parameter) {
 			return readValue(object, parameter.getType(), parent);
 		}
 	}
+
+	/**
+	 * Take a generated id that was set on the target, and set it on the source that is returned to the caller.
+	 * 1) id is a String and the source has a setId(String id) method.
+	 * 2) <S extends T> S id -> source.setId(T)
+	 * 3) X id -> source.setId(Y.newInstance(X))
+	 * 4) X id -> source.setId(Y.from(X));
+	 * 5) X id -> source.setId(Y.newInstance(X.toString()) this will only work if T.toString() is appropriate
+	 * 6) X id -> source.setId(Y.fromString(X))  this will only work if T.toString() is appropriate
+	 * If none of that works, try to set the id or @Id field directly
+	 *
+	 * @param source object being converted
+	 * @param id     id that was used in the target
+	 */
+	private static void setId(Object source, Object id) {
```

Comment:
That's not how it's supposed to work. Spring Data provides a `PersistentEntity`/`PersistentProperty` meta-model to identify ID properties. Property values are read and written through `PersistentPropertyAccessor` that ensure proper handling of immutable objects (that is creating a new instance for Kotlin data classes and returning the modified instance).